File: osprey/include/omp/omp_lib.f
!  OpenMP runtime library to be used in conjunction with Open64 Compiler Suites.
!
!  Copyright (C) 2003 - 2009 Tsinghua University.
!
!  This library is free software; you can redistribute it and/or
!  modify it under the terms of the GNU Lesser General Public
!  License as published by the Free Software Foundation; either
!  version 2.1 of the License, or (at your option) any later version.
!
!  This library is distributed in the hope that it will be useful,
!  but WITHOUT ANY WARRANTY; without even the implied warranty of
!  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
!  Lesser General Public License for more details.
!
!  You should have received a copy of the GNU Lesser General Public
!  License along with this library; if not, write to the Free Software
!  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
!  
!  Contact information: HPC Institute, Department of Computer Science and Technology,
!  Tsinghua University, Beijing 100084, CHINA, or:
!
!  http://hpc.cs.tsinghua.edu.cn

        module omp_lib_kinds
        integer, parameter :: omp_integer_kind   = 4
        integer, parameter :: omp_logical_kind   = 4
        integer, parameter :: omp_lock_kind      = 8
        integer, parameter :: omp_nest_lock_kind = 8
        end module omp_lib_kinds

        module omp_lib
          use omp_lib_kinds
          integer, parameter :: openmp_version = 199910

          interface
            subroutine omp_destroy_lock (var)
              use omp_lib_kinds
              integer (kind=omp_lock_kind), intent(inout) :: var
            end subroutine omp_destroy_lock
          end interface

          interface
            subroutine omp_destroy_nest_lock (var)
              use omp_lib_kinds
              integer (kind=omp_nest_lock_kind), intent(inout) :: var
            end subroutine omp_destroy_nest_lock
          end interface

          interface
            function omp_get_dynamic ()
              use omp_lib_kinds
              logical (kind=omp_logical_kind) :: omp_get_dynamic
            end function omp_get_dynamic
          end interface

          interface
            function omp_get_max_threads ()
              use omp_lib_kinds
              integer (kind=omp_integer_kind) :: omp_get_max_threads
            end function omp_get_max_threads
          end interface

          interface
            function omp_get_nested ()
              use omp_lib_kinds
              logical (kind=omp_logical_kind) :: omp_get_nested
            end function omp_get_nested
          end interface

          interface
            function omp_get_num_procs ()
              use omp_lib_kinds
              integer (kind=omp_integer_kind) :: omp_get_num_procs
            end function omp_get_num_procs
          end interface

          interface
            function omp_get_num_threads ()
              use omp_lib_kinds
              integer (kind=omp_integer_kind) :: omp_get_num_threads
            end function omp_get_num_threads
          end interface

          interface
            function omp_get_thread_num ()
              use omp_lib_kinds
              integer (kind=omp_integer_kind) :: omp_get_thread_num
            end function omp_get_thread_num
          end interface

          interface
            function omp_get_wtick ()
              use omp_lib_kinds
              double precision :: omp_get_wtick
            end function omp_get_wtick
          end interface

          interface
            function omp_get_wtime ()
              use omp_lib_kinds
              double precision :: omp_get_wtime
            end function omp_get_wtime
          end interface

          interface
            subroutine omp_init_lock (var)
              use omp_lib_kinds
              integer (kind=omp_lock_kind), intent(out) :: var
            end subroutine omp_init_lock
          end interface

          interface
            subroutine omp_init_nest_lock (var)
              use omp_lib_kinds
              integer (kind=omp_nest_lock_kind), intent(out) :: var
            end subroutine omp_init_nest_lock
          end interface

          interface
            function omp_in_parallel ()
              use omp_lib_kinds
              logical (kind=omp_logical_kind) :: omp_in_parallel
            end function omp_in_parallel
          end interface

          interface
            subroutine omp_set_dynamic (enable)
              use omp_lib_kinds
              logical (kind=omp_logical_kind), intent(in) :: enable
            end subroutine omp_set_dynamic
          end interface

          interface
            subroutine omp_set_lock (var)
              use omp_lib_kinds
              integer (kind=omp_lock_kind), intent(inout) :: var
            end subroutine omp_set_lock
          end interface

          interface
            subroutine omp_set_nest_lock (var)
              use omp_lib_kinds
              integer (kind=omp_nest_lock_kind), intent(inout) :: var
            end subroutine omp_set_nest_lock
          end interface

          interface
            subroutine omp_set_nested (enable)
              use omp_lib_kinds
              logical (kind=omp_logical_kind), intent(in) :: enable
            end subroutine omp_set_nested
          end interface

          interface
            subroutine omp_set_num_threads (nthreads)
              use omp_lib_kinds
              integer (kind=omp_integer_kind), intent(in) :: nthreads
            end subroutine omp_set_num_threads
          end interface

          interface
            function omp_test_lock (var)
              use omp_lib_kinds
              logical (kind=omp_logical_kind) :: omp_test_lock
              integer (kind=omp_lock_kind), intent(inout) :: var
            end function omp_test_lock
          end interface

          interface
            function omp_test_nest_lock (var)
              use omp_lib_kinds
              integer (kind=omp_integer_kind) :: omp_test_nest_lock
              integer (kind=omp_nest_lock_kind), intent(inout) :: var
            end function omp_test_nest_lock
          end interface

          interface
            subroutine omp_unset_lock (var)
              use omp_lib_kinds
              integer (kind=omp_lock_kind), intent(inout) :: var
            end subroutine omp_unset_lock
          end interface

          interface
            subroutine omp_unset_nest_lock (var)
              use omp_lib_kinds
              integer (kind=omp_nest_lock_kind), intent(inout) :: var
            end subroutine omp_unset_nest_lock
          end interface
        end module omp_lib

File: osprey/libopenmp/omp_lib.c
/*
 * Copyright (C) 2009 Advanced Micro Devices, Inc.  All Rights Reserved.
 */

/*

  OpenMP runtime library to be used in conjunction with Open64 Compiler Suites.

  Copyright (C) 2003 - 2009 Tsinghua University.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
  
  Contact information: HPC Institute, Department of Computer Science and Technology,
  Tsinghua University, Beijing 100084, CHINA, or:

  http://hpc.cs.tsinghua.edu.cn
  
*/

/*
 * File: omp_lib.c
 * Abstract: implementation of OpenMP run-time library subroutines
 *          for OpenMP programmer
 * History: 04/23/2003, built by Jiang Hongshan, Tsinghua Univ.
 * 
 */

#include <stdlib.h>

#include "omp_rtl.h"
#include "omp_lock.h" 
#include "omp_util.h"

#include <sys/time.h>

void
__omp_fatal(char *msg)
{
  fprintf(stderr, "OpenMP Run-time Library FATAL error:\n %s\n", msg);
  exit(-1);
}


/*
 * OpenMP standard library function
 */

void
omp_set_num_threads(omp_int_t num)
{
  __ompc_set_num_threads(num);
}

void omp_set_num_threads_(omp_int_t * num)
{
    __ompc_set_num_threads(*num);
}

omp_int_t
omp_get_num_threads(void)
{
  return (omp_int_t)__ompc_get_num_threads();
}

omp_int_t omp_get_num_threads_(void);
#pragma weak omp_get_num_threads_ = omp_get_num_threads

omp_int_t
omp_get_max_threads(void)
{
  return (omp_int_t)__ompc_get_max_threads();
}


omp_int_t omp_get_max_threads_(void);
#pragma weak omp_get_max_threads_ = omp_get_max_threads

omp_int_t
omp_get_thread_num(void)
{
  return (omp_int_t)__ompc_get_local_thread_num();
}


omp_int_t omp_get_thread_num_(void);
#pragma weak omp_get_thread_num_ = omp_get_thread_num

omp_int_t
omp_get_num_procs(void)
{
  return (omp_int_t)__ompc_get_num_procs();
}


omp_int_t omp_get_num_procs_(void);
#pragma weak omp_get_num_procs_ = omp_get_num_procs
omp_int_t
omp_in_parallel(void)
{
  return (omp_int_t)__ompc_in_parallel();
}


omp_int_t omp_in_parallel_(void);
#pragma weak omp_in_parallel_ = omp_in_parallel

void
omp_set_dynamic(omp_int_t dynamic)
{
  __ompc_set_dynamic(dynamic);
}


void omp_set_dynamic_(omp_int_t* dynamic)
{
    __ompc_set_dynamic(*dynamic);
}

omp_int_t
omp_get_dynamic(void)
{
  return (omp_int_t)__ompc_get_dynamic();
}


omp_int_t omp_get_dynamic_(void);
#pragma weak omp_get_dynamic_ = omp_get_dynamic

void
omp_set_nested(omp_int_t nested)
{
  __ompc_set_nested(nested);
}


void omp_set_nested_(omp_int_t* nested)
{
    __ompc_set_nested(*nested);
}

omp_int_t
omp_get_nested(void)
{
  return (omp_int_t)__ompc_get_nested();
}

omp_int_t omp_get_nested_(void);
#pragma weak omp_get_nested_ = omp_get_nested
/*
 * Lock Functions
 */
extern int __omp_spin_user_lock;
void
omp_init_lock(volatile omp_lock_t *lock)
{
  // put the lock aligned to cache line size and
  // malloc a cache line for the lock to avoid
  // false sharing
  ompc_lock_t *tmp_lp;
  tmp_lp = aligned_malloc(sizeof(ompc_lock_t), CACHE_LINE_SIZE);
  Is_True(tmp_lp != NULL, "can not allocate tmp_lp");

  tmp_lp->flag = __omp_spin_user_lock;
  __ompc_init_lock(tmp_lp);
  (*lock) = (omp_lock_t)tmp_lp; 
}


void omp_init_lock_(volatile omp_lock_t *);
#pragma weak omp_init_lock_ = omp_init_lock

void
omp_init_nest_lock(volatile omp_nest_lock_t *lock)
{
  // put the lock aligned to cache line size
  ompc_nest_lock_t * tmp_lp;
  tmp_lp = aligned_malloc(sizeof(ompc_nest_lock_t), CACHE_LINE_SIZE); 
  Is_True(tmp_lp != NULL, "can not allocate tmp_lp");
  __ompc_init_nest_lock(tmp_lp);
  (*lock) = (omp_nest_lock_t)tmp_lp; 
}


void omp_init_nest_lock_(volatile omp_nest_lock_t *);
#pragma weak omp_init_nest_lock_ = omp_init_nest_lock

void
omp_destroy_lock(volatile omp_lock_t *lock)
{
  __ompc_destroy_lock((ompc_lock_t*)(*lock));
  aligned_free((ompc_lock_t*)(*lock)); 
}


void omp_destroy_lock_(volatile omp_lock_t *);
#pragma weak omp_destroy_lock_ = omp_destroy_lock
void
omp_destroy_nest_lock(volatile omp_nest_lock_t *lock)
{
  __ompc_destroy_nest_lock((ompc_nest_lock_t*)(*lock));
  aligned_free((ompc_nest_lock_t*)(*lock)); 
}


void omp_destroy_nest_lock_(volatile omp_nest_lock_t *);
#pragma weak omp_destroy_nest_lock_ = omp_destroy_nest_lock
void
omp_set_lock(volatile omp_lock_t *lock)
{
  __ompc_lock((ompc_lock_t*)(*lock));
}


void omp_set_lock_(volatile omp_lock_t *);
#pragma weak omp_set_lock_ = omp_set_lock
void
omp_set_nest_lock(volatile omp_nest_lock_t *lock)
{
  __ompc_nest_lock((ompc_nest_lock_t*)(*lock));
}


void omp_set_nest_lock_(volatile omp_nest_lock_t *);
#pragma weak omp_set_nest_lock_ = omp_set_nest_lock
void
omp_unset_lock(volatile omp_lock_t *lock)
{
  __ompc_unlock((ompc_lock_t*)(*lock));
}


void omp_unset_lock_(volatile omp_lock_t *);
#pragma weak omp_unset_lock_ = omp_unset_lock
void
omp_unset_nest_lock(volatile omp_nest_lock_t *lock)
{
  __ompc_nest_unlock((ompc_nest_lock_t*)(*lock));
}


void omp_unset_nest_lock_(volatile omp_nest_lock_t *);
#pragma weak omp_unset_nest_lock_ = omp_unset_nest_lock
int
omp_test_lock(volatile omp_lock_t *lock)
{
  return __ompc_test_lock((ompc_lock_t*)(*lock));
}


int omp_test_lock_(volatile omp_lock_t *);
#pragma weak omp_test_lock_ = omp_test_lock
int
omp_test_nest_lock(volatile omp_nest_lock_t *lock){
  return __ompc_test_nest_lock((ompc_nest_lock_t*)(*lock));
}
	

int omp_test_nest_lock_(volatile omp_nest_lock_t *);
#pragma weak omp_test_nest_lock_ = omp_test_nest_lock
/*
 * Timer function
 */
omp_wtime_t
omp_get_wtime(void)
{
  struct timeval tval;

  gettimeofday(&tval, NULL);
  return (omp_wtime_t)( (double)tval.tv_sec + 1.0e-6 * (double)tval.tv_usec );
}

omp_wtime_t omp_get_wtime_(void);
#pragma weak omp_get_wtime_ = omp_get_wtime


omp_wtime_t omp_get_wtime__(void);
#pragma weak omp_get_wtime__ = omp_get_wtime

omp_wtime_t
omp_get_wtick(void)
{
  double t1, t2;
	
  t1 = omp_get_wtime();
  do {
    t2 = omp_get_wtime();
  } while(t1 == t2);

  return (omp_wtime_t)(t2 - t1);
}


omp_wtime_t omp_get_wtick_(void);
#pragma weak omp_get_wtick_ = omp_get_wtick


omp_wtime_t omp_get_wtick__(void);
#pragma weak omp_get_wtick__ = omp_get_wtick
